In checkpoints, serialize hash of attachment instead of its data

andr3ej opened the pull request:
This PR implements https://r3-cev.atlassian.net/browse/CORDA-300

mikehearn commented:
I'm not totally sure why a new interface is required? Can't we just change the attachment storage so it does its own dbtx? or, run checkpoint deserialisation inside a transaction in the SMM?

Using the serialize-as-token infrastructure might not be the best way to do this, if it has to be extended. It's really meant for singleton classes. Not something like an attachment.

rick-r3 commented:
There should be a database transaction _in-flight_ when restoring the checkpoints since they themselves are read from the database, so.... agree we shouldn't need a new interface (just use `AttachmentStorage`) or to open any new database transactions.

Well done for finding `SerializeAsToken` but, as Mike is saying, it wasn't designed for this since it assumes a finite set of tokens (see the presence of an internal map) which isn't the case here.  I think you'll end up having to write a custom serialiser in `Kryo.kt`.

andr3ej commented:
Changes since the last version of this PR:
* Quasar won't let us have a reference to the db at deserialisation time (it clears all ThreadLocals) so attachments now reload their data lazily
* SerializeAsToken with custom tokens appears to do exactly what I want so I'm not keen to reimplement that, it's only SingletonSerializationToken that backs itself by the singleton map...
* ... which I've refactored so we can get singletons out by supertype, which AbstractAttachment does to get the StorageService for reloading attachment data (practical motivation is it can't see the StorageServiceImpl class)

mikehearn commented on core/src/main/kotlin/net/corda/core/contracts/Structures.kt:
```diff
@@ -513,3 +512,21 @@ interface Attachment : NamedByHash {
     }
 }
+
+abstract class AbstractAttachment(dataLoader: () -> ByteArray) : Attachment {
```
Is this really meant to be in the core module? I'd think it'd be an implementation detail of the node.

andr3ej replied:
It's used by FetchAttachmentsFlow which is in core

andr3ej commented on core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt:
```diff
@@ -75,14 +82,14 @@ class SerializeAsTokenSerializer<T : SerializeAsToken> : Serializer<T>() {
 class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool) {
-    internal val tokenToTokenized = mutableMapOf<SerializationToken, SerializeAsToken>()
-    internal var readOnly = false
+    private val classNameToSingletons = Multimaps.newMultimap<String, SerializeAsToken>(HashMap(), { newSetFromMap(IdentityHashMap()) })
```
Yes, the changes are essentially 2 things:
* Add uniqueSingletonImplementing and related data structure changes
* Code moved from the token to the context as more things can be private that way

I think making the SATC more specific would work, but:
* That would be a bigger-impact change for another PR, I haven't actually changed what's stored in the SATC in this PR, only how it's stored
* This approach (lookup by type, analogous to constructor injection) means you don't have to remember a chain of field names to get to the object you want, see Law of Demeter
* I'm preparing a PR to remove Kryo.attachmentStorage that will take advantage of this facility

rick-r3 replied:
I think adding `ServiceHub` to SATC is a smaller change.  It's just the constructor needs an extra `val services: ServiceHub`, and then change the single call site (outside tests).   No other changes to STAC would then be required. Whilst you may have a point re: Law of Demeter I'd defer that to how `ServiceHub` looks and is implemented, although personally I prefer my services to be discoverable via code complete in the IDE...

andr3ej replied:
Done

andr3ej commented on core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt:
```diff
@@ -95,30 +102,44 @@ class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool) {
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun <T : Any> uniqueSingletonImplementing(clazz: KClass<T>) = uniqueSingleton(clazz.java.name, false) as T
```
Done

rick-r3 commented on core/src/main/kotlin/net/corda/flows/FetchAttachmentsFlow.kt:
```diff
@@ -15,19 +18,21 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
-        override fun equals(other: Any?) = other === this || other is Attachment && other.id == this.id
-        override fun hashCode(): Int = id.hashCode()
-        override fun toString(): String = "${javaClass.simpleName}(id=$id)"
+    private class ByteArrayAttachment(dataLoader: () -> ByteArray) : AbstractAttachment(dataLoader), SerializeAsToken {
```
I wonder if this should be named `ImportedAttachment` or `PersistedAttachment` or something like that?  `ByteArrayAttachment` sounds like it is standalone, but in fact for the tokenisation to work it must be already persisted (which it always is for this flow).

rick-r3 commented on core/src/main/kotlin/net/corda/core/serialization/SerializationToken.kt:
```diff
@@ -74,15 +76,15 @@ class SerializeAsTokenSerializer<T : SerializeAsToken> : Serializer<T>() {
-    internal val tokenToTokenized = mutableMapOf<SerializationToken, SerializeAsToken>()
-    internal var readOnly = false
+class SerializeAsTokenContext(toBeTokenized: Any, kryoPool: KryoPool, val serviceHub: ServiceHub) {
+    private val classNameToSingleton = mutableMapOf<String, SerializeAsToken>()
```
We could have left this alone, but agree this is a fair change to discourage accidental caching of non-singletons that might bloat out the map, so let's go with it.

rick-r3 commented on core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt:
```diff
@@ -0,0 +1,192 @@
+}
+
+class AttachmentSerializationTest {
+
```
Remove blank line

rick-r3 commented on core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt:
```diff
@@ -0,0 +1,192 @@
+    private class ClientResult(internal val attachmentContent: String)
+
+    private abstract class ClientLogic(server: MockNetwork.MockNode) : FlowLogic<ClientResult>() {
+
```
Remove blank line

rick-r3 commented on core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt:
```diff
@@ -0,0 +1,192 @@
+
+        @Suspendable // This annotation is required by the instrumentation verifier.
+        internal abstract fun getAttachmentContent(): String
+
```
Remove blank line

rick-r3 commented on core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt:
```diff
@@ -0,0 +1,192 @@
+        client.hackAttachment(attachmentId, "hacked")
+        assertEquals("hacked", rebootClientAndGetAttachmentContent(false))
+    }
+
```
Remove blank line

andr3ej commented:
Changes made
